docs/book/command-params.md

```diff
@@ -0,0 +1,378 @@
+# Command Params
+
+The Symfony Console component allows commands to define
+[input arguments and options](https://symfony.com/doc/current/console/input.html).
+
+The laminas-cli package adds a third possibility, **input parameters**, which
+allow you to define command options that will become interactive prompts when
+omitted during command invocation.
+
+## Using input parameters
+
+Internally, input parameters behave like standard options with the following
+modifications:
+
+- Input parameters are optional by default.
+- In interactive mode, when retrieving a parameter value that was not provided
+  as an option, the user will be asked to provide the value.
+- In non-interactive mode, if the user does not supply the option during
+  invocation, and the parameter is not marked as required, its
+  default value will be used; otherwise, an exception will be thrown,
+- **If the parameter has validations or normalizations supplied, these will be
+  applied regardless of whether the option is provided during invocation or when
+  the user is prompted to provide it.**
+- If the user provides a value via a prompt, that value will be stored as an
+  input option so it can be passed to the [next command in the
+  chain](commands-chain.md) (if one is defined).
+
+## Input parameter types
+
+Input parameters are classes that implement
+`Laminas\Cli\Input\InputParamInterface`:
+
+```php
+namespace Laminas\Cli\Input;
+
+use Symfony\Component\Console\Question\Question;
+
+interface InputParamInterface
+{
+    /**
+     * Default value to use if none provided.
+     *
+     * @return null|mixed
+     */
+    public function getDefault();
+
+    public function getDescription(): string;
+
+    public function getName(): string;
+
+    /**
+     * Return the Symfony\Component\Console\Input\InputOption::VALUE_* type.
+     */
+    public function getOptionMode(): ?int;
+
+    public function getShortcut(): ?string;
+
+    public function getQuestion(): Question;
+
+    public function isRequired(): bool;
+
+    /**
+     * @param mixed $defaultValue
+     */
+    public function setDefault($defaultValue): self;
```

Comment:
Could we split this into `Capable` and `Aware` interfaces?
Only that class which creates the instance needs the "setters" while the consumer classes only need the "getters".

I dunno why, but I prefer having these decoupled. Thoughts?

Reply:
The consumer of the getters is the `ParamAwareCommandTrait`.

The consumer of the setters is the user creating commands and parameters.

In the majority of use cases, you will either use one of the parameter types we provide — in which case you will *definitely* need to customize things like the description and default value — or you will create one on-the-fly as an anonymous class. Even in this latter scenario, it will be easier to code the anonymous class and use the defined setter methods than it would be to create a constructor that sets and/or accepts defaults.

Basically, I think that splitting the interface into two would offer correctness at the expense of the most common use cases. As a consumer, I would expect that I can call the base setter methods on _any_ parameter instance, and would likely be caught by surprise if I tried and got a "method not defined" error.